### ISSUE TITLE
Not getting all VPN clients reported from Checkpoint firewall

### ISSUE THREAD

**Issue by jezekus** [librenms/librenms#13058]
Problem with loading all VPN clients via SNMP on Checkpoint when following IP address is lower than previos one.
Snmpwalk expect inceasing sequence and as part of OID is remote peer IP.

**Versions:**

- LibreNMS 21.7.0
- OS version: CentOS 7.9
- SNMP version: 5.7.8
- Checkpoint: R80.30

**Detail:**
- snmpbulkwalk to check checkpoint VPNs, IP address of remote peer is part of OID
`snmpbulkwalk -v 2c -c 12345678  10.20.30.40 .1.3.6.1.4.1.2620.500.9002.1.3`

- Error: OID not increasing: SNMPv2-SMI::enterprises.2620.500.9002.1.3.172.30.195.12.0
` >= SNMPv2-SMI::enterprises.2620.500.9002.1.3.172.30.122.99.0`

 - fix for snmpbulkwalk use in command -Cc
`snmpbulkwalk -v 2c -Cc -c 12345678  10.20.30.40 .1.3.6.1.4.1.2620.500.9002.1.3`

- Probably need to fix in discovery too During PHP discovery got error message
`php discovery.php -h 10.20.30.40 -d -m sensors`

- Error: OID not increasing: tunnelPeerObjName.172.30.195.12.0
`>= tunnelPeerObjName.172.30.122.99.0`

**Comment by rhinoau** [comment]
I don't have any checkpoint experience and early days with me on OS support - but trying to define this a bit more to assist.

The test data is in order to such a level it looks like the device is meant to return it as such.

- May this be a bug with the checkpoint release?
- -or- is it just pure luck the test data came out in order?
- Do you have any other devices to compare against to confirm?

I'm assuming this is gaia, you can add this line to the includes/definitions/discovery/gaia.yaml file under the tunnelState block to see if it resolves it for discovery, then you'll need to test if it holds up for polling too:

> `snmp_flags: ['-OQUs', '-Cc']`

The first bunch of settings are a common default but all the different discovery bits are different code.  Run discovery with debug (-d) to see what it is actually trying it with, and make it the same. If that all works, a fix won't be hard.

**Comment by jezekus** [comment]
This helped a bit. All tunnels were added but there are still problems with names and IP addresses. In DB is () as a name because tunnelPeerIpAddr and  tunnelPeerObjName are checked without -Cc parameter.

 

## Tunnel state

SNMP[[0;36m'/usr/bin/snmpbulkwalk' '-v2c' '-c' 'COMMUNITY' '-OQUs' '-Cc' '-Ih' '-m' 'CHECKPOINT-MIB' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/checkpoint' 'udp:HOSTNAME:161'

 

## Tunnel IP address

Caching data: os sensors SNMP[[0;36m'/usr/bin/snmpbulkwalk' '-v2c' '-c' 'COMMUNITY' '-OteQUs' '-Ih' '-m' 'CHECKPOINT-MIB' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/checkpoint' 'udp:HOSTNAME:161' 'tunnelPeerIpAddr'[0m]

Exitcode: 1 

Error: OID not increasing: tunnelPeerIpAddr.10.20.30.1.0

>= tunnelPeerIpAddr.10.20.22.1.0

 

## Tunnel Objecct name

SNMP[[0;36m'/usr/bin/snmpbulkwalk' '-v2c' '-c' 'COMMUNITY' '-OteQUs' '-Ih' '-m' 'CHECKPOINT-MIB' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/checkpoint' 'udp:HOSTNAME:161' 'tunnelPeerObjName'[0m]

Exitcode: 1 

Error: OID not increasing: tunnelPeerObjName.10.20.30.1.0

>= tunnelPeerObjName.10.20.22.1.0

**Comment by rhinoau** [comment]
> Tunnel IP address
> 
> Caching data: os sensors SNMP[[0;36m'/usr/bin/snmpbulkwalk' '-v2c' '-c' 'COMMUNITY' '-OteQUs' '-Ih' '-m' 'CHECKPOINT-MIB' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/checkpoint' 'udp:HOSTNAME:161' 'tunnelPeerIpAdd
> 

Not in a position to test the actual data - but you can put the Cc snmp_flags in the pre-cache like below, confirmed it works and parses at a basic level my side:

```
    sensors:
        pre-cache:
            data:
                -
                    oid:
                        - tunnelPeerIpAddr
                        - tunnelPeerObjName
                        - tempertureSensorName
                        - fanSpeedSensorName
                        - voltageSensorName
                        - fwLSConnName
                        - mglsGWIP
                    snmp_flags: ['-OQUs', '-Cc']
```
If you do that also, does it fix that issue?

**Comment by jezekus** [comment]
Thank you for the suggestion.
I have just tested it and it works!

Would it be possible to merge it into master?

**Comment by murrant** [comment]
@jezekus just need someone to send a pull request with the change.

Also, FYI, OIDs must only increase according to the standard.  So the device is spitting out invalid data and this is actually net-snmp throwing the error, not LibreNMS.   Might be a good idea to raise a bug with the vendor as they incorrectly sorting this data before sending it to snmp clients.  (also could be duplicate oid, which very bad)

**Comment by jezekus** [comment]
@murrant  understand, I will create PR with a change and also discuss and report bug report to Checkpoint.

**Comment by laf** [comment]
This was fixed in #15878
